receiver/receivercreator/resourceenhancer.go

```diff
@@ -0,0 +1,77 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+var _ consumer.MetricsConsumer = (*resourceEnhancer)(nil)
+
+// resourceEnhancer adds additional resource attribute entries
+// from the given endpoint environment. The added attributes vary based on the type
+// of the endpoint.
+type resourceEnhancer struct {
+	nextConsumer consumer.MetricsConsumer
```

Comment:
Can we find a way not to make resourceEnhancer to be only possible to use with metrics? We don't really care about metrics, we only work with the resource.

Reply:
See above, I don't think so given how the consumers are modeled. There's no generic `consumer.Consumer` interface for instance.

Reply:
resourceEnhancer does not have to be a composable MetricConsumer, does it? Only because it is it has to implement metrics specific interface.
Up to you, but I think we are making it more difficult to add non-metric types to receiver_creator in the future.

Reply:
I see. I'm not too worried about it. I think it'll be trivial to do the necessary refactoring if it needs to support `ConsumeLogs` one day. I just tried factoring out the common logic but it was like 4 lines of code because you still have to have the telemetry type-specific code to iterate through, e.g.:

```go
	rm := md.ResourceMetrics()
	for i := 0; i < rm.Len(); i++ {
		rms := rm.At(i)
		attrs := rms.Resource().Attributes()

                // This is the only common code.
		for attr, val := range r.attrs {
			attrs.InsertString(attr, val)
		}
	}
```